### PR TITLE
Refactor transact() related methods.

### PR DESCRIPTION
**Pull request by jianglai**
This PR makes a few changes to make it possible to turn on
  per-transaction isolation level with minimal disruption:

  1) Changed the signatures of transact() and reTransact() methods to allow
  passing in lambdas that throw checked exceptions. Previously one has
  always to wrap such lambdas in try-and-retrow blocks, which wasn't a
  big issue when one can liberally open nested transactions around small
  lambdas and keeps the "throwing" part outside the lambda. This becomes a
  much bigger hassle when the goal is to eliminate nested transactions and
  put as much code as possible within the top-level lambda. As a result,
  the transactNoRetry() method now handles checked exceptions by re-throwing
  them as runtime exceptions.

  2) Changed the name and meaning of the config file field that used to
  indicate if per-transaction isolation level is enabled or not. Now it
  decides if transact() is called within a transaction, whether to
  throw or to log, regardless whether the transaction could have
  succeeded based on the isolation override level (if provided). The
  flag will initially be set to false and would help us identify all
  instances of nested calls and either refactor them or use reTransact()
  instead. Once we are fairly certain that no nested calls to transact()
  exists, we flip the flag to true and start enforcing this logic.
  Eventually the flag will go away and nested calls to transact() will
  always throw.

  3) Per-transaction isolation level will now always be applied, if an
  override is provided. Because currently there should be no actual
  use of such feature (except for places where we explicitly use an
  override and have ensured no nested transactions exist, like in
  RefreshDnsForAllDomainsAction), we do not expect any issues with
  conflicting isolation levels, which would resulted in failure.

  3) transactNoRetry() is made package private and removed from the
  exposed API of JpaTransactionManager. This saves a lot of redundant
  methods that do not have a practical use. The only instances where this
  method was called outside the package was in the reader of
  RegistryJpaIO, which should have no problem with retrying.

- - -
This change is [Reviewable](https://reviewable.io/reviews/google/nomulus/2195)

### CODE REVIEW

**Review by jianglai — COMMENTED** [review]
> Reviewed 11 of 11 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 1 unresolved discussion (waiting on @github-advanced-security[bot])

___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 155 at r1](https://reviewable.io/reviews/google/nomulus/2195#gh-1373654274:-NhhNAwQ1tqjfC6imhFN:b-896fix) ([raw file](https://github.com/google/nomulus/blob/889eaa79ac9cd871924b26a77ddd19d4b6d201f9/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L155)):*
<details><summary><i>Previously, github-advanced-security[bot] wrote…</i></summary><blockquote>

## Missing space in string literal

This string appears to be missing a space after 'nested'.

[Show more details](https://github.com/google/nomulus/security/code-scanning/1577)
</blockquote></details>

Done.

**Review by jianglai — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 1 unresolved discussion (waiting on @github-advanced-security[bot])

**Review by jianglai — COMMENTED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 1 unresolved discussion (waiting on @CydeWeys, @github-advanced-security[bot], and @weiminyu)

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 13 unresolved discussions (waiting on @gbrodman, @github-advanced-security[bot], @jianglai, and @weiminyu)

___
*[`core/src/main/java/google/registry/config/RegistryConfigSettings.java` line 116 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAmPyG-hWk3NxOH0ym:-NiAmPyG-hWk3NxOH0yn:b-hqo4dz) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/config/RegistryConfigSettings.java#L116)):*
> ```Java
>   /** Configuration for Hibernate. */
>   public static class Hibernate {
>     public boolean disableNestedTransactions;
> ```

Generally config settings should be phrased in the positive, rather than in the negative. It can be hard to parse a false value for something that is itself already a negation.  See http://go/tott-530

I think `allowNestedTransactions` makes sense.
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 153 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiApra64WZkJ0JJBRG_:-NiApra64WZkJ0JJBRGa:b2tmn8t) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L153)):*
> ```Java
>   public <T> T transact(Callable<T> work, TransactionIsolationLevel isolationLevel) {
>     if (inTransaction()) {
>       String message =
> ```

This should be a `static final String` on the class.
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 159 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAq-0dEg0zAaW6DwY-:-NiAq-0dEg0zAaW6DwY0:b-ltdnzn) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L159)):*
> ```Java
>         throw new IllegalStateException(message);
>       }
>       logger.atWarning().log(message);
> ```

Without logging a stack trace or something, is this actually all that actionable?
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 191 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAoR918fe1dE36dbWj:-NiAoR918fe1dE36dbWk:b-x5ut33) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L191)):*
> ```Java
>       } catch (Exception | Error e) {
>         tryWrapAndThrow(e);
>         if (e instanceof RuntimeException) {
> ```

Better to use the recommended new practice at https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 210 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiApjGNA08Lmg7OC85F:-NiApjGNA08Lmg7OC85G:bhcodis) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L210)):*
> ```Java
>             .executeUpdate();
>         logger.atInfo().log(
>             "Override transaction isolation level from %s to %s",
> ```

Should be `Overridding`
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 225 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAoT7OAjrymVRcTMv4:-NiAoT7OAjrymVRcTMv5:b-n8re1i) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L225)):*
> ```Java
>       }
>       tryWrapAndThrow(e);
>       if (e instanceof RuntimeException) {
> ```

Ditto on https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 227 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAn1PCA5W3jmLO8Wls:-NiAn1PCA5W3jmLO8Wlt:b-4y09q2) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L227)):*
> ```Java
>    */
>   @FunctionalInterface
>   interface ThrowableRunnable {
> ```

I think `ThrowingRunnable` makes more sense from an English grammar perspective and overall Java ecosystem consistency perspective, a la https://junit.org/junit4/javadoc/4.13/org/junit/function/ThrowingRunnable.html

To me a "throwable runnable" could somehow mean a runnable that is thrown, rather than a runnable that throws.
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 227 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAnvAT-dh_fnnL3TLd:-NiAnvAU791Wouh58taJ:b-3pgwed) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L227)):*
> ```Java
>    */
>   @FunctionalInterface
>   interface ThrowableRunnable {
> ```

Should this be public? Or will it never be referred to from outside of the transaction manager classes themselves?
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 227 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAnwDp6mMZSxVHjtL2:-NiAnwDp6mMZSxVHjtL3:b-uk0e4o) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L227)):*
> ```Java
>    */
>   @FunctionalInterface
>   interface ThrowableRunnable {
> ```

How is this different from a `Callable<Void>`?
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 228 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAmt4A9MWmtVtrnA1H:-NiAmt4A9MWmtVtrnA1I:b-si140n) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L228)):*
> ```Java
>   @FunctionalInterface
>   interface ThrowableRunnable {
>     void run() throws Exception;
> ```

Should this be `throws Throwable`?
___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 105 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAo_KJ0efe6RpczQqQ:-NiAo_KJ0efe6RpczQqR:bfmc9vs) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L105)):*
> ```Java
>       try {
>         return work.call();
>       } catch (Exception | Error e) {
> ```

Seems like this should just be `catch (Throwable e)`
___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 106 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAod4X7dY8o76Di847:-NiAod4X7dY8o76Di848:b-m3d8u9) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L106)):*
> ```Java
>         return work.call();
>       } catch (Exception | Error e) {
>         tryWrapAndThrow(e);
> ```

It seems confusing to have both this and then also the subsequent 5 lines.  But if you turn it into:

```
  tryWrapAndThrow(e);
  throwIfUnchecked(e);
  throw new RuntimeException(e);
```

I think it becomes more clear.

Also, the naming of `tryWrapAndThrow()` is pretty generic, and doesn't actually say much about what it's doing. A name like `throwIfSqlException()` or `throwIfDatabaseException()` would make it more clear.

**Review by jianglai — COMMENTED** [review]
> Reviewed 10 of 10 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 13 unresolved discussions (waiting on @CydeWeys, @gbrodman, @github-advanced-security[bot], and @weiminyu)

___
*[`core/src/main/java/google/registry/config/RegistryConfigSettings.java` line 116 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAmPyG-hWk3NxOH0ym:-NiG7NiQ3AjyR9Bhswzr:b-896fix) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/config/RegistryConfigSettings.java#L116)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Generally config settings should be phrased in the positive, rather than in the negative. It can be hard to parse a false value for something that is itself already a negation.  See http://go/tott-530

I think `allowNestedTransactions` makes sense.
</blockquote></details>

Done.
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 153 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiApra64WZkJ0JJBRG_:-NiG7e822QWbIIM_sENz:b-896fix) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L153)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

This should be a `static final String` on the class.
</blockquote></details>

Done.
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 159 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAq-0dEg0zAaW6DwY-:-NiG82vh9KfvLERN2I3e:b-oz8tyz) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L159)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Without logging a stack trace or something, is this actually all that actionable?
</blockquote></details>

Good call!
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 191 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAoR918fe1dE36dbWj:-NiGGirA1-hUsyeD1Rfs:b-896fix) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L191)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Better to use the recommended new practice at https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate
</blockquote></details>

Done.
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 210 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiApjGNA08Lmg7OC85F:-NiG8Dox0vVb4ZGngSt-:b-896fix) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L210)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Should be `Overridding`
</blockquote></details>

Done.
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 225 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAoT7OAjrymVRcTMv4:-NiGGjXx6OJQL0TWyL1J:b-896fix) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L225)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Ditto on https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate
</blockquote></details>

Done.
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 227 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAn1PCA5W3jmLO8Wls:-NiG8MJVDLDpXZxJvxnF:b3qvaua) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L227)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

I think `ThrowingRunnable` makes more sense from an English grammar perspective and overall Java ecosystem consistency perspective, a la https://junit.org/junit4/javadoc/4.13/org/junit/function/ThrowingRunnable.html

To me a "throwable runnable" could somehow mean a runnable that is thrown, rather than a runnable that throws.
</blockquote></details>

Haha it looks like non-throwing runnables are a pain for not just us.
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 227 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAnvAT-dh_fnnL3TLd:-NiG8cIx3kcxkXegqW8j:by2mzag) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L227)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Should this be public? Or will it never be referred to from outside of the transaction manager classes themselves?
</blockquote></details>

I don't envision it every being used explicitly. Anyone wanting to pass a lambda to one of the transct methods should always just use the `() -> {}` notation, instead of creating an anonymous class.
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 227 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAnwDp6mMZSxVHjtL2:-NiG8v7BADJdJFqpvFm7:bg7idez) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L227)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

How is this different from a `Callable<Void>`?
</blockquote></details>

The main advantage is that you don't have to `return null`. Makes the call site less bloated.
___
*[`core/src/main/java/google/registry/persistence/transaction/TransactionManager.java` line 228 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAmt4A9MWmtVtrnA1H:-NiG9LvaBFwzNiiIgpR6:b-h717jy) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/main/java/google/registry/persistence/transaction/TransactionManager.java#L228)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Should this be `throws Throwable`?
</blockquote></details>

It seems like use code should always throw `Exception`s (which include all the checked exceptions that we expect to throw):

https://stackoverflow.com/questions/498217/when-should-throwable-be-used-instead-of-new-exception
___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 105 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAo_KJ0efe6RpczQqQ:-NiGFhKXAUVakD4OMDbo:b-896fix) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L105)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Seems like this should just be `catch (Throwable e)`
</blockquote></details>

Done.
___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 106 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAod4X7dY8o76Di847:-NiGGgDpDYbc3DKRO5Ew:bvelciq) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L106)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

It seems confusing to have both this and then also the subsequent 5 lines.  But if you turn it into:

```
  tryWrapAndThrow(e);
  throwIfUnchecked(e);
  throw new RuntimeException(e);
```

I think it becomes more clear.

Also, the naming of `tryWrapAndThrow()` is pretty generic, and doesn't actually say much about what it's doing. A name like `throwIfSqlException()` or `throwIfDatabaseException()` would make it more clear.
</blockquote></details>

This does look much better.

**Review by gbrodman — DISMISSED** [review]
> Reviewed 3 of 11 files at r1, 9 of 10 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 15 unresolved discussions (waiting on @CydeWeys, @github-advanced-security[bot], @jianglai, and @weiminyu)

___
*[`core/src/main/java/google/registry/config/files/default-config.yaml` line 192 at r6](https://reviewable.io/reviews/google/nomulus/2195#-NiGT0xX8jBI_v-cIZBr:-NiGT0xX8jBI_v-cIZBs:b-ziiqoz) ([raw file](https://github.com/google/nomulus/blob/a760bdbd9ad7d98db59565dcd5bbc734f96ac68e/core/src/main/java/google/registry/config/files/default-config.yaml#L192)):*
> ```YAML
> 
> hibernate:
>   # If set to false, calls to tm().transact() cannot be nested. If set to false,
> ```

second "false" should be true, and the "N" that starts the next line should be lowercase 
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 191 at r6](https://reviewable.io/reviews/google/nomulus/2195#-NiGUh6YEmHCAKLO4lk4:-NiGUh6YEmHCAKLO4lk5:b-zgblno) ([raw file](https://github.com/google/nomulus/blob/a760bdbd9ad7d98db59565dcd5bbc734f96ac68e/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L191)):*
> ```Java
>       try {
>         return work.call();
>       } catch (Throwable e) {
> ```

Same thing here, shouldn't we catch Exception? 
___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 105 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAo_KJ0efe6RpczQqQ:-NiGU0EX1Y29-uNgA72w:b82l682) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L105)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Shouldn't we catch Exception here similar to what we do elsewhere in the PR? I can't imagine that we would want to handle `Error`s at this level

**Review by jianglai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 15 unresolved discussions (waiting on @CydeWeys, @gbrodman, @github-advanced-security[bot], and @weiminyu)

___
*[`core/src/main/java/google/registry/config/files/default-config.yaml` line 192 at r6](https://reviewable.io/reviews/google/nomulus/2195#-NiGT0xX8jBI_v-cIZBr:-NiGbQO0-dx-hUkCLCHQ:b-896fix) ([raw file](https://github.com/google/nomulus/blob/a760bdbd9ad7d98db59565dcd5bbc734f96ac68e/core/src/main/java/google/registry/config/files/default-config.yaml#L192)):*
<details><summary><i>Previously, gbrodman wrote…</i></summary><blockquote>

second "false" should be true, and the "N" that starts the next line should be lowercase
</blockquote></details>

Done.
___
*[`core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java` line 191 at r6](https://reviewable.io/reviews/google/nomulus/2195#-NiGUh6YEmHCAKLO4lk4:-NiGc7q13Jslr-WO6G9-:b1druz3) ([raw file](https://github.com/google/nomulus/blob/a760bdbd9ad7d98db59565dcd5bbc734f96ac68e/core/src/main/java/google/registry/persistence/transaction/JpaTransactionManagerImpl.java#L191)):*
<details><summary><i>Previously, gbrodman wrote…</i></summary><blockquote>

Same thing here, shouldn't we catch Exception?
</blockquote></details>

Ditto, errors will be re-thrown (because they are unchecked), as if they were not caught.
___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 105 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAo_KJ0efe6RpczQqQ:-NiGbm8FAx4_5WohR_LN:b-p602vh) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L105)):*
<details><summary><i>Previously, gbrodman wrote…</i></summary><blockquote>

Shouldn't we catch Exception here similar to what we do elsewhere in the PR? I can't imagine that we would want to handle `Error`s at this level
</blockquote></details>

The existing logic catches both Exceptions and Errors. Errors will just be re-thrown. I think in practice it's probably fine here to not handle Errors, but I don't want to touch things what was working fine before.

**Review by jianglai — COMMENTED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 15 unresolved discussions (waiting on @CydeWeys, @gbrodman, @github-advanced-security[bot], and @weiminyu)

**Review by CydeWeys — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 6 unresolved discussions (waiting on @gbrodman, @github-advanced-security[bot], @jianglai, and @weiminyu)

___
*[`core/src/main/java/google/registry/config/RegistryConfig.java` line 1551 at r7](https://reviewable.io/reviews/google/nomulus/2195#-NiLN_br0xzbVdCxKxC8:-NiLN_br0xzbVdCxKxC9:b-3tillt) ([raw file](https://github.com/google/nomulus/blob/d905c2d810b4ef6884b35bf51c09036ec1018c72/core/src/main/java/google/registry/config/RegistryConfig.java#L1551)):*
> ```Java
>   }
> 
>   /** Returns true if nested calls to {@code tm().transact()} is allowed. */
> ```

s/is/are/

**Review by CydeWeys — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 4 unresolved discussions (waiting on @gbrodman, @github-advanced-security[bot], @jianglai, and @weiminyu)

___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 105 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAo_KJ0efe6RpczQqQ:-NiLOC_O3QDlvG7YyqlT:be6h7sk) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L105)):*
<details><summary><i>Previously, jianglai (Lai Jiang) wrote…</i></summary><blockquote>

The existing logic catches both Exceptions and Errors. Errors will just be re-thrown. I think in practice it's probably fine here to not handle Errors, but I don't want to touch things what was working fine before.
</blockquote></details>

Based on the link you included above, it seems like it would make sense to not catch `Error`. To the extent we were doing that before, it was probably unnecessary?

**Review by jianglai — COMMENTED** [review]
> Reviewed 2 of 3 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 4 unresolved discussions (waiting on @CydeWeys, @gbrodman, @github-advanced-security[bot], and @weiminyu)

___
*[`core/src/main/java/google/registry/config/RegistryConfig.java` line 1551 at r7](https://reviewable.io/reviews/google/nomulus/2195#-NiLN_br0xzbVdCxKxC8:-NiLflmX8qWGJXKgmxf1:b-896fix) ([raw file](https://github.com/google/nomulus/blob/d905c2d810b4ef6884b35bf51c09036ec1018c72/core/src/main/java/google/registry/config/RegistryConfig.java#L1551)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

s/is/are/
</blockquote></details>

Done.
___
*[`core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java` line 105 at r5](https://reviewable.io/reviews/google/nomulus/2195#-NiAo_KJ0efe6RpczQqQ:-NiLdzm07CMazLahRGas:bjflquv) ([raw file](https://github.com/google/nomulus/blob/86cf4f698b44cea802c9486ff4ef61ae6ea363a3/core/src/test/java/google/registry/persistence/transaction/ReplicaSimulatingJpaTransactionManager.java#L105)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Based on the link you included above, it seems like it would make sense to not catch `Error`. To the extent we were doing that before, it was probably unnecessary?
</blockquote></details>

I think the only place where catching a `Throwable` is necessary is when actually executing the query, because we want to attempt a rollback if any failure arises, including `Errors`.

**Review by CydeWeys — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 4 unresolved discussions (waiting on @gbrodman, @github-advanced-security[bot], @jianglai, and @weiminyu)

___
*[`core/src/test/java/google/registry/persistence/transaction/JpaTransactionManagerImplTest.java` line 787 at r9](https://reviewable.io/reviews/google/nomulus/2195#-NiLgIZJ2sAvJbvT1ONP:-NiLgIZK9irkNfeO2m_q:b-l8rxnn) ([raw file](https://github.com/google/nomulus/blob/135d370cb3fd9d076f0bf3baaf7ff7c3404eed09/core/src/test/java/google/registry/persistence/transaction/JpaTransactionManagerImplTest.java#L787)):*
> ```Java
>     tm().assertInTransaction();
>     TransactionIsolationLevel currentLevel = tm().getCurrentTransactionIsolationLevel();
>     if (currentLevel != expectedLevel) {
> ```

Seems like a good use case for `checkState()`

**Review by jianglai — COMMENTED** [review]
> Reviewed 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/google/nomulus/2195)* status: all files reviewed, 4 unresolved discussions (waiting on @CydeWeys, @gbrodman, @github-advanced-security[bot], and @weiminyu)

___
*[`core/src/test/java/google/registry/persistence/transaction/JpaTransactionManagerImplTest.java` line 787 at r9](https://reviewable.io/reviews/google/nomulus/2195#-NiLgIZJ2sAvJbvT1ONP:-NiLwugn3onGVfRo_FvA:b-896fix) ([raw file](https://github.com/google/nomulus/blob/3db886b534fce5e3bca14049cf7670b6140bc385/core/src/test/java/google/registry/persistence/transaction/JpaTransactionManagerImplTest.java#L787)):*
<details><summary><i>Previously, CydeWeys (Ben McIlwain) wrote…</i></summary><blockquote>

Seems like a good use case for `checkState()`
</blockquote></details>

Done.
